employee_scheduling.py

```diff
@@ -218,3 +199,144 @@ def run_cqm(cqm):
         return sampleset, errors
 
     return feasible_sampleset, None
+
+
+def build_nl(params: ModelParams) -> tuple[Model, BinaryVariable]:
+    """Builds an employee scheduling nonlinear model.
+
+    Args:
+        params (ModelParams): model parameters
+
+    Returns:
+        tuple[Model, BinaryVariable]: the NL model and assignments decision variable
+    """
+    # Create list of employees
+    employees = list(params.availability.keys())
+    model = Model()
+
+    # Create a binary symbol representing the assignment of employees to shifts
+    # i.e. assignments[employee][shift] = 1 if assigned, else 0
+    num_employees = len(employees)
+    num_shifts = len(params.shifts)
+    assignments = model.binary((num_employees, num_shifts))
+
+    # Create availability constant
+    availability_list = [params.availability[employee] for employee in employees]
+    for i, sublist in enumerate(availability_list):
+        availability_list[i] = [a if a != 2 else 100 for a in sublist]
```

Comment:
In the context of the Dash app, where $a_{e,s}$ is employee $e$'s availability for shift $s$, a value of $a_{e,s}=0$ is assigned if the employee is unavailable for that shift, $a_{e,s}=1$ if they are available, and $a_{e,s}=2$ if the employee prefers that shift. These are the values in the assignments parameter that is passed into the model-building function. The objective for the model is then calculated as $\sum_e^{|E|}\sum_s^{|S|}a_{e,s}x_{e,s}$, where $x_{e,s}$ is the decision variable representing employee $e$'s assignment to shift $s$. In my experimentation, I found that boosting the value of $a_{e,s}$ to 100 from 2 helped the NL solver find better solutions faster since assignment of a preferred shift would have a greater impact on the objective value. The exact value of 100 is somewhat arbitrary though, I just found that a value of 10 didn't help the speedup much and 1,000 performed about as well as 100.

Reply:
Added a comment to add context to this